numpy/_core/src/umath/dispatching.c

```diff
@@ -964,11 +965,20 @@ promote_and_get_ufuncimpl(PyUFuncObject *ufunc,
              */
             Py_CLEAR(op_dtypes[i]);
         }
+        /*
+         * If the op_dtype ends up being a non-legacy one, then we cannot use
+         * legacy promotion (unless this is a python scalar).
+         */
+        if (op_dtypes[i] != NULL && !NPY_DT_is_legacy(op_dtypes[i]) && (
+                signature[i] != NULL ||  // signature cannot be a pyscalar
+                !(PyArray_FLAGS(ops[i]) & NPY_ARRAY_WAS_PYTHON_LITERAL))) {
+            allow_legacy_promotion = NPY_FALSE;
```

Comment:
Might it be better to call this variable "legacy_promotion_is_possible"? (no big deal, obviously, but since it is now an internal variable, might as well make it as clear as possible)